File: php/logbrew-php/tests/support_ticket.php
<?php

declare(strict_types=1);

use LogBrew\SupportTicketDraft;

$supportDraft = SupportTicketDraft::create(
    source: 'sdk',
    category: 'ingest_failure',
    title: '  PHP ingest failed  ',
    description: '  Local support draft for an explicit user handoff.  ',
    projectId: 'proj_public_123',
    environment: 'production',
    runtime: PHP_VERSION,
    framework: 'laravel',
    sdkPackage: 'logbrew/sdk',
    sdkVersion: '0.1.0',
    release: 'checkout@1.2.3',
    traceId: '4BF92F3577B34DA6A3CE929D0E0E4736',
    eventId: 'evt_issue_001',
    diagnostics: [
        'authorization' => 'Bearer lbw_ingest_secret_value',
        'endpoint' => 'https://api.example.com/v1/events?token=secret#fragment',
        'localPath' => '/Users/example/project/.env',
        'debugNote' => 'failed at https://api.example.com/v1/events?token=secret from /Users/example/project/.env',
        'runtime' => PHP_VERSION,
        'attempts' => [1, 2, 3],
        'exception' => new RuntimeException('do not include this message'),
        'nested' => [
            'cookie' => 'session=secret',
            'tokenText' => 'token=secret',
            'safe' => 'kept',
        ],
        'objectValue' => new stdClass(),
        'nan' => NAN,
    ]
);
assertTrue($supportDraft['source'] === 'sdk', 'expected support draft source');
assertTrue($supportDraft['category'] === 'ingest_failure', 'expected support draft category');
assertTrue($supportDraft['title'] === 'PHP ingest failed', 'expected support draft trimmed title');
assertTrue($supportDraft['description'] === 'Local support draft for an explicit user handoff.', 'expected support draft trimmed description');
assertTrue($supportDraft['project_id'] === 'proj_public_123', 'expected support draft project id');
assertTrue($supportDraft['sdk_package'] === 'logbrew/sdk', 'expected support draft package');
assertTrue($supportDraft['trace_id'] === '4bf92f3577b34da6a3ce929d0e0e4736', 'expected support draft normalized trace id');
$supportDiagnostics = $supportDraft['diagnostics'] ?? null;
if (!is_array($supportDiagnostics)) {
    fwrite(STDERR, 'expected support draft diagnostics object' . PHP_EOL);
    exit(1);
}
assertTrue(($supportDiagnostics['authorization'] ?? null) === '[redacted]', 'expected support draft authorization redaction');
assertTrue(($supportDiagnostics['endpoint'] ?? null) === '[redacted-url]/v1/events', 'expected support draft URL redaction');
assertTrue(($supportDiagnostics['localPath'] ?? null) === '[redacted-path]', 'expected support draft local path redaction');
assertTrue(($supportDiagnostics['debugNote'] ?? null) === 'failed at [redacted-url]/v1/events from [redacted-path]', 'expected support draft embedded URL and path redaction');
$supportException = $supportDiagnostics['exception'] ?? null;
if (!is_array($supportException)) {
    fwrite(STDERR, 'expected support draft exception object' . PHP_EOL);
    exit(1);
}
assertTrue(($supportException['type'] ?? null) === 'RuntimeException', 'expected support draft exception type only');
$supportNested = $supportDiagnostics['nested'] ?? null;
if (!is_array($supportNested)) {
    fwrite(STDERR, 'expected support draft nested diagnostics object' . PHP_EOL);
    exit(1);
}
assertTrue(($supportNested['cookie'] ?? null) === '[redacted]', 'expected support draft nested cookie redaction');
assertTrue(($supportNested['tokenText'] ?? null) === '[redacted]', 'expected support draft nested token text redaction');
assertTrue(($supportNested['safe'] ?? null) === 'kept', 'expected support draft safe nested value');
$supportJson = json_encode($supportDraft, JSON_THROW_ON_ERROR);
foreach ([
    'lbw_ingest_secret_value',
    'api.example.com',
    'token=secret',
    '/Users/example/project',
    'do not include this message',
    'objectValue',
    'nan',
] as $needle) {
    assertTrue(!str_contains($supportJson, $needle), "support draft leaked diagnostic value: {$needle}");
}
expectThrows(
    fn () => SupportTicketDraft::create(
        source: 'bot',
        category: 'ingest_failure',
        title: 'Bad source',
        description: 'Bad source'
    ),
    'support ticket source must be one of'
);
expectThrows(
    fn () => SupportTicketDraft::create(
        source: 'sdk',
        category: 'ingest_failure',
        title: 'Bad trace',
        description: 'Bad trace',
        traceId: '00000000000000000000000000000000'
    ),
    'support ticket trace_id must be 32 non-zero hex characters'
);
expectThrows(
    fn () => SupportTicketDraft::create(
        source: 'sdk',
        category: 'not_real',
        title: 'Bad category',
        description: 'Bad category'
    ),
    'support ticket category must be one of'
);

File: ruby/logbrew-ruby/lib/logbrew/support_ticket.rb
# frozen_string_literal: true

module LogBrew
  # Local-only support-ticket payload drafts for explicit user or agent handoff.
  module SupportTicketDraft
    SOURCES = %w[cli sdk website docs mobile].freeze
    CATEGORIES = %w[
      sdk_install_failure
      ingest_failure
      auth_failure
      project_setup
      dashboard_issue
      docs_confusion
      cli_issue
      mobile_issue
      billing_question
      other
    ].freeze
    SENSITIVE_KEY_MARKERS = %w[
      apikey
      auth
      authorization
      authtoken
      bearer
      clientsecret
      connectionstring
      cookie
      credential
      dsn
      email
      errormessage
      exceptionmessage
      password
      passwd
      privatekey
      refreshtoken
      secret
      session
      setcookie
      stacktrace
      token
      traceback
    ].freeze
    REDACTED = "[redacted]"
    MAX_DEPTH = 5
    MAX_ARRAY_LENGTH = 20
    MAX_STRING_LENGTH = 500

    module_function

    def create(source:, category:, title:, description:, **options)
      Validation.require_allowed_value("support ticket source", source, SOURCES)
      Validation.require_allowed_value("support ticket category", category, CATEGORIES)

      draft = {
        "source" => source,
        "category" => category,
        "title" => required_string("support ticket title", title),
        "description" => required_string("support ticket description", description)
      }
      add_optional_string(draft, "project_id", options[:project_id])
      add_optional_string(draft, "environment", options[:environment])
      add_optional_string(draft, "runtime", options[:runtime])
      add_optional_string(draft, "framework", options[:framework])
      add_optional_string(draft, "sdk_package", options[:sdk_package])
      add_optional_string(draft, "sdk_version", options[:sdk_version])
      add_optional_string(draft, "release", options[:release])
      add_trace_id(draft, options[:trace_id])
      add_optional_string(draft, "event_id", options[:event_id])
      add_diagnostics(draft, options[:diagnostics]) if options.key?(:diagnostics)
      draft
    end

    def required_string(label, value)
      Validation.require_non_empty(label, value)
      value.strip
    end
    private_class_method :required_string

    def add_optional_string(draft, key, value)
      return if value.nil?

      draft[key] = required_string("support ticket #{key}", value)
    end
    private_class_method :add_optional_string

    def add_trace_id(draft, value)
      return if value.nil?

      normalized = value.to_s.strip.downcase
      unless normalized.match?(/\A[0-9a-f]{32}\z/) && !normalized.delete("0").empty?
        raise SdkError.new("validation_error", "support ticket trace_id must be 32 non-zero hex characters")
      end
      draft["trace_id"] = normalized
    end
    private_class_method :add_trace_id

    def add_diagnostics(draft, diagnostics)
      unless diagnostics.is_a?(Hash)
        raise SdkError.new("validation_error", "support ticket diagnostics must be an object")
      end

      sanitized = sanitize_diagnostic_hash(diagnostics, 0)
      draft["diagnostics"] = sanitized unless sanitized.empty?
    end
    private_class_method :add_diagnostics

    def sanitize_diagnostic_hash(value, depth)
      value.each_with_object({}) do |(key, child), safe|
        break safe if safe.length >= MAX_ARRAY_LENGTH

        normalized_key = key.to_s
        next if normalized_key.strip.empty?

        sanitized = if sensitive_key?(normalized_key)
                      REDACTED
                    else
                      sanitize_diagnostic_value(child, depth + 1)
                    end
        safe[normalized_key] = sanitized unless sanitized.nil?
      end
    end
    private_class_method :sanitize_diagnostic_hash

    def sanitize_diagnostic_value(value, depth)
      return { "type" => value.class.name } if value.is_a?(Exception)
      return nil if depth > MAX_DEPTH

      case value
      when nil, true, false, String, Integer
        sanitize_primitive(value)
      when Float
        value.finite? ? value : nil
      when Hash
        sanitize_diagnostic_hash(value, depth)
      when Array
        value.first(MAX_ARRAY_LENGTH).map { |item| sanitize_diagnostic_value(item, depth + 1) }.compact
      else
        nil
      end
    end
    private_class_method :sanitize_diagnostic_value

    def sanitize_primitive(value)
      return value unless value.is_a?(String)

      text = value.strip
      return "" if text.empty?
      return REDACTED if sensitive_string?(text)

      redacted = redact_url(redact_local_path(text))
      redacted.length > MAX_STRING_LENGTH ? "#{redacted[0, MAX_STRING_LENGTH]}..." : redacted
    end
    private_class_method :sanitize_primitive

    def sensitive_key?(key)
      normalized = key.downcase.gsub(/[^a-z0-9]/, "")
      SENSITIVE_KEY_MARKERS.any? { |marker| normalized.include?(marker) }
    end
    private_class_method :sensitive_key?

    def sensitive_string?(value)
      value.match?(/(?:authorization|api[_-]?key|token|secret|password|passwd|cookie)\s*[:=]/i) ||
        value.match?(/\bBearer\s+[A-Za-z0-9._~+\/=-]+/i) ||
        value.match?(/\blbw_(?:ingest|client|api)_[A-Za-z0-9._-]+/i) ||
        value.match?(/\b(?:github_pat|ghp|gho|npm|pypi|sk_live|sk_test|xox[baprs]|AKIA)[A-Za-z0-9._-]+/)
    end
    private_class_method :sensitive_string?

    def redact_local_path(value)
      return "[redacted-path]" if value.match?(/\A(?:\/Users\/|\/home\/|\/var\/folders\/|[A-Za-z]:\\)/)

      value
    end
    private_class_method :redact_local_path

    def redact_url(value)
      uri = URI.parse(value)
      return "[redacted-url]#{uri.path.empty? ? '/' : uri.path}" if uri.is_a?(URI::HTTP) && uri.host

      value.split(/[?#]/, 2)[0]
    rescue URI::InvalidURIError
      value.split(/[?#]/, 2)[0]
    end
    private_class_method :redact_url
  end
end
